Scripted streams

Jacob-Mango opened the pull request.

Jacob-Mango committed:
moved XML to Scripts

Jacob-Mango committed:
XML code cleanup

Jacob-Mango committed:
Added 'CF_StringStream' and 'CF_BinaryStream'

Jacob-Mango committed:
Added 'byte' and 'char', fixed infinite reading in CF_BinaryStream

Jacob-Mango committed:
Changed 'ReadFile' and 'WriteFile' to 'File'

Jacob-Mango committed:
renamed common variable to index in CF_StringStream::File

Jacob-Mango committed:
Fixed writing null bytes

Jacob-Mango committed:
Added reinterpret casting

Jacob-Mango committed:
Read and write types for binary stream, removed typedefs

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/Stream/CF_BinaryStream.c:
```diff
@@ -0,0 +1,286 @@
+		Write(0);
+	}
+
+	override int Read()
```
Read ... what? A byte, if so this should be ReadByte()

Otherwise causes confusion with the other ReadX methods.

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Cast.c:
```diff
@@ -0,0 +1,13 @@
+
+		return dst_arr[0];
+	}
+};
```
New line at the end of files please.

Jacob-Mango replied:
Weird, it has a new line for me on my PC. Git isn't detecting new file changes.

Arkensor replied:
I do not think git is to blame here, even the addition of an empty line would be a stageable change

Arkensor replied:
Only Github shows it like this ... when cloned there is a newline.

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/Stream/CF_BinaryStream.c:
```diff
@@ -0,0 +1,286 @@
+			{
+				int lastZero = 0;
+
+				bool varr[1];
```
More descriptive name for `varr`

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/Stream/CF_BinaryStream.c:
```diff
@@ -0,0 +1,286 @@
+		return m_Data.Count();
+	}
+
+	override void GoTo(int position)
```
Instead of GoTo Use Seek with an origin enum. 

s. below

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/Stream/CF_BinaryStream.c:
```diff
@@ -0,0 +1,286 @@
+		m_Position = position;
+	}
+
+	override void Seek(int offset)
```
Prototype should be `void Seek(int offset, CF_Stream_Origin origin)` where origin is an Enum with for example `SEEK_SET`, `SEEK_CUR`, `SEEK_END` ... inspired by [c++ fseek](https://www.cplusplus.com/reference/cstdio/fseek/)

Default value for origin should be `SEEK_CUR`

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/Stream/CF_Stream.c:
```diff
@@ -0,0 +1,183 @@
+		/*CF.Log.*/Error("Not implemented.");
+	}
+
+	void SetBytes(array<int> bytes)
```
Why is there a SetBytes method? Normally streams are only interacted with via Read and Write and the only full data operation is GetBytes?

Jacob-Mango replied:
Syncing over the network through RPC (although inefficient).

Generally, people shouldn't use it but giving options is better than nothing

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/XML/CF_XML_Callback.c:
```diff
@@ -0,0 +1,10 @@
+typedef CF_XML_Callback XMLCallback;
```
Not sure why git did not detect this as a file location move. Please validate that there were no accidental changes.

Arkensor commented on JM/CF/Scripts/3_Game/CommunityFramework/XML/CF_XML.c:
```diff
@@ -4,7 +4,7 @@ class CF_XML
 	{
 	}
 
-	static ref CF_XML_Reader Open(string path)
```
Are you 100% sure that **ALL** the refs should be removed? (In this file and the other XML related ones)
If so good, otherwise please check if they were needed somewhere.

Arkensor reviewed (changes requested):
A few minor questions and suggestions. Please check all of them before requesting a re-review.

Jacob-Mango committed:
Resolved PR discussions

Arkensor committed:
Fixed formatting

Jacob-Mango committed:
Documentation

Arkensor commented on docs/Streams/index.md:
```diff
@@ -0,0 +1,83 @@
+if (stream.File("$profile:Test.png"))
+{
+  array<int> data = stream.GetBytes();
+  GetRPCManager().
```
I'd leave RPC out of this tbh. Also, this line is incomplete regardless.

Arkensor commented on docs/Streams/index.md:
````diff
@@ -0,0 +1,83 @@
+
+```
+CF_StringStream stream();
+stream.File("$profile:test.txt", FileMode.READ);
````
Maybe have a look at c++ fstream. It had an optional constructor parameter where you could put in the file and the mode already which internally called Open

I think binarystream and stringstream are ok, but if you want to interact with files, then that should be a CF_Filestream ... which you could open from one of the other streams to get their data into the file or read from the file again. E.g.

```cs
CF_StringStream data();
data.Write("Hello Jacob");
CF_FileStream fileWrite("<path>", Filemode.OPTION);
fileWrite.Write(data);
fileWrite.Close();
```

Your examples all show file interaction ... maybe you should first demonstrate the streams for what they are named after, aka reading and writing strings or binary data ... and the use-case together with files should be coming a bit later - or be its own kind of stream.

Arkensor reviewed (commented):
Conceptual suggestions

Jacob-Mango committed:
added base byte types

Jacob-Mango committed:
Compile error fix

Jacob-Mango committed:
- Renamed Write to WriteByte
- ReadByte and WriteByte now use CF_Byte
- GetBytes and SetBytes now use array<CF_Byte>

Jacob-Mango committed:
Updated documentation

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Operations.c:
```diff
@@ -0,0 +1,8 @@
+	int c = ~y;
+	int d = (b | c);
+	return (a & d);
+}
```
Missing newline

Arkensor commented on docs/Streams/index.md:
```diff
@@ -8,76 +8,79 @@ There are currently two Streams avaliable, `CF_BinaryStream` and `CF_StringStrea
 if (stream.File("$profile:Test.png", FileMode.WRITE))
 {
-   stream.SetBytes(data);
+  stream.SetBytes(data);
```
Wrong indent

Jacob-Mango committed:
renamed global ^ to CF_XOR, added duplicate XOR to CF_Uint

Jacob-Mango committed:
Updated index.md

Jacob-Mango committed:
split streams into writers and readers

Jacob-Mango committed:
Renamed CF_File to CF_IO

Jacob-Mango committed:
Added CF_PackedByte

Jacob-Mango committed:
Added CF_Stream

Jacob-Mango committed:
Moved IO to 1_Core

Jacob-Mango committed:
Added String and Serializer streams

Jacob-Mango committed:
Merge branch 'development' into feature/serializer

Jacob-Mango committed:
- Added Hex Stream
- Fixed String Stream
- Added more Encoding functions
- Added read and write primitive methods for TextWriter/TextReader
- Made Stream optional parameter for IO
- Added ReadWord and ReadWhitespace to TextReader

Jacob-Mango committed:
- Fixed seeking and writing
- Optimized binary
- Added close method

Jacob-Mango committed:
Fixed EOF next and previous

Jacob-Mango committed:
Compile error fixes

Jacob-Mango committed:
Serializer stream fixes

Jacob-Mango committed:
Documentation

Jacob-Mango committed:
Added output to hex

Jacob-Mango committed:
Updated streams header

Jacob-Mango committed:
Updated documentation

Jacob-Mango committed:
Added CF_SHA256

Jacob-Mango committed:
Renamed CF_HexStream to CF_Base16Stream

Jacob-Mango committed:
Fixed 'CopyTo' and added 'CF_Base64Stream'

Jacob-Mango committed:
Added 'CopyTo' with size

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Byte.c:
```diff
@@ -0,0 +1,73 @@
+	{
+		CF_Byte old_byte = value;
+		value = value & 255;
+		if (value != old_byte)
```
Remove snake case.

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Byte.c:
```diff
@@ -0,0 +1,73 @@
+		return CF_Encoding.BASE_16[n0] + CF_Encoding.BASE_16[n1];
+	}
+
+	override CF_Uint ^(CF_Uint y)
+	{
+		CF_Byte b = super.^(y);
+		b.Clamp();
+		return b;
+	}
```
Are we really keeping this? I know it is cool at a first glance but in code ehhh ... I think XOR written out below is enough.

Jacob-Mango replied:
Yes

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_PackedByte.c:
```diff
@@ -0,0 +1,117 @@
+class CF_PackedByte
```
What is a packed byte? Does this mean bit stuffing as in multiple values in one byte?

Jacob-Mango replied:
Linked list of `CF_Byte`, optimized reading and writing.

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/IO/CF_Base16Stream.c:
```diff
@@ -0,0 +1,110 @@
+	override string ToStr()
+	{
+		return super.ToStr();
+		//return Encode();
```
remove unused code

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/IO/CF_IO.c:
```diff
@@ -0,0 +1,208 @@
+class CF_IO : Managed
```
I understand the idea of a base stream, and it is good, but I would add one abstraction layer in between, to make sure you do not come to the idea to Read on a write stream or vice-versa.
Just CF_IO should not allow to read and write, one has to decide what they want it to be by strong typing it to something higher.

Move WriteX into inherited CF_WriteStream
Move ReadX into a CF_ReadStream

And inherit the other readers and writers from those intermediate classes.

Jacob-Mango replied:
Doesn't solve anything by having an additional layer of abstraction, just makes it harder to re-use the same variable name in the same scope.

Arkensor replied:
This solves not knowing what interaction is allowed on a given stream. You are not supposed to first use a stream instance to read and then turn it into a writer. Those are usually two separate streams, of two separate types. Reading only on a ReadStream and Writing only on a WriteStream.

Jacob-Mango replied:
I will add a return method for `IsRead` and `IsWrite`. 

There may be a use case in the future where an InputOutput operation can be done at the same time. Rather not lock CF into a forced inheritance.

Arkensor commented on docs/Streams/index.md:
```diff
@@ -0,0 +1,73 @@
+# Streams
+
+A stream is a doubly linked list of bytes. It can be used for file input/output, DayZ serializers, hex conversion or string parsing. Once a stream is finished, it is recommended to close the stream either by letting the reference count equal zero or call the `Close` method.
```
Once you do not longer need the stream, it is...

Jacob-Mango replied:
I don't understand your statement

Arkensor replied:
Your wording just sounds odd. Replace "Once a stream is finished" with what I said/rewrite it. A stream is not finished - it is not being used anymore or empty.

Arkensor commented on docs/Streams/index.md:
````diff
@@ -0,0 +1,73 @@
+CF_BinaryReader reader = new CF_BinaryReader(new CF_SerializerReadStream(ctx.GetReadContext()));
+Print(reader.ReadFloat());	// "0.1"
+reader.Close();
+```
````
Missing example for SHA256 ...

Arkensor commented on docs/Streams/index.md:
````diff
@@ -0,0 +1,73 @@
+```csharp
+string path = "$profile:test.txt";
+
+if (FileExist(path)) DeleteFile(path);
````
The file mode should determine if the file is replaced or appended, so this line has no real value does it? I would remove it.

Arkensor commented on docs/Streams/index.md:
```diff
@@ -0,0 +1,73 @@
+CF_HexStream stream = new CF_HexStream();
+
+CF_BinaryWriter writer = new CF_BinaryWriter(stream);
+writer.WriteFloat(5);
+writer.Close();
+
+Print(stream.ToStr());	// "40A00000"
```
If you make an example about float, maybe append the endianness and float format to the comment.

Arkensor reviewed (changes requested):
Overall very solid, if it is all tested and works as intended then this will be a great feature. Only a few change requests, the rest is questions or suggestions.

I thought about if you could add a `string CF_Encoding.ToHex(int value)` and `int CF_Encoding.FromHex(string value)` shortcuts so one must not go through the steam overhead.

Jacob-Mango committed:
Snek case

Jacob-Mango committed:
Removed unused code

Jacob-Mango committed:
Added short int support

Jacob-Mango committed:
- Documented endianess, fixed all to little-endian (most common in x86)
- Documented SHA-256

Jacob-Mango committed:
Addressing PR changes

Jacob-Mango committed:
Fixed crash when `Close` isn't called.

Jacob-Mango committed:
Compile fix with 1.14

Jacob-Mango commented:
> I thought about if you could add a string CF_Encoding.ToHex(int value) and int CF_Encoding.FromHex(string value) shortcuts so one must not go through the steam overhead.

This already exists in a slightly different syntax.

To Hex:

```cpp
CF_Byte value = 80;
string hex = value.ToHex();
```

```cpp
array<CF_Byte> bytes = new array<CF_Byte>();
bytes.Insert(80);
string hex = CF_Encoding.ToHex(bytes);
```

From Hex:

```cpp
string hex = "FF";
array<CF_Byte> bytes = CF_Encoding.FromHex(hex);
```

Jacob-Mango committed:
Removed DAYZ_1_14 compile defines, added 'GetDebugName' to 'CF_IO'

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Byte.c:
```diff
@@ -0,0 +1,62 @@
+	CF_Byte Clamp()
+	{
+		CF_Byte oldByte = value;
+		value = value & 255;
+		return value;
```
Why not 
```
return (value & 255);
```
to save the assignment

Or is this about in-place modification and returning the value is just for convinience?

Jacob-Mango replied:
> Or is this about in-place modification and returning the value is just for convenience?

Correct, so both `cfbyte.Clamp()` and `x = cfbyte.Clamp()` compile

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Byte.c:
```diff
@@ -0,0 +1,62 @@
+	void Set(int index, int v)
+	{
+		value = v;
+		value.Clamp();
```
It might be significantly faster to not make Clamp a function and instead have `& 255` in the code each time.

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Long.c:
```diff
@@ -0,0 +1,12 @@
+/*
+class CF_Long
+{
+	int I0;
+	int I1;
+
+	string GetDebugName()
+	{
+		return "" + I0 + I1;
+	}
+};
+*/
```
Unused or discarded concept? Then delete the file, please.

Arkensor commented on JM/CF/Scripts/1_Core/CommunityFramework/CF_Uint.c:
```diff
@@ -0,0 +1,138 @@
+		CF_Uint a = value;
+		CF_Uint carry = a & b;
+		CF_Uint result = a.XOR(b);
+		while (carry != 0)
```
You can just do `while(carry)` because any value other than 0 is implicitly true. Saves the comparison on each iteration.

Arkensor commented on docs/Cryptography/index.md:
```diff
@@ -0,0 +1,21 @@
+string path = "$profile:test.txt";
+
+// Create the file
+CF_TextReader reader = new CF_TextReader(new CF_FileStream(path, FileMode.READ));
```
rename reader to `input`

Arkensor commented on docs/Cryptography/index.md:
```diff
@@ -0,0 +1,21 @@
+CF_TextReader reader = new CF_TextReader(new CF_FileStream(path, FileMode.READ));
+Print(reader.ReadLine()); // "This is some text"
+
+CF_Base16Stream hexStream = new CF_Base16Stream();
```
rename hexStream to `output`

That makes it easier to understand what is what in the `Process` call

Arkensor commented on docs/Streams/index.md:
````diff
@@ -0,0 +1,73 @@
+
+Use the `CF_StringStream` class to read and write to a string source. This stream does not accept '\0' bytes.
+
+```csharp
+string data = "Hello, World!";
+
+CF_TextReader reader = new CF_TextReader(new CF_StringStream(data));
+Print("ReadWord: " + reader.ReadWord());				// "Hello"
+Print("ReadChar: " + reader.ReadChar());				// ","
+Print("ReadWhitespace: " + reader.ReadWhitespace());	// ""
+Print("ReadWord: " + reader.ReadWord());				// "World"
+Print("ReadChar: " + reader.ReadChar());				// "!"
+reader.Close();
+```
````
Maybe you could give a bit of information on how we detect what a word is? I would assume it reads until the next separator char? If so a quick list of which chars for this would be nice. Or if it's the rule "characters that are not aA-zZ" then say that.

Arkensor commented on docs/Streams/index.md:
```diff
@@ -0,0 +1,73 @@
+
+## Serializer
+
+Use the `CF_SerializerReadStream` class to read to from the DayZ serializer and `CF_SerializerWriteStream` to write to the DayZ serializer.
```
I would clarify that it can interact with anything that inherits from `Serializer`, otherwise one could think you only mean ScriptReadWriteContext. 

> Use `CF_SerializerReadStream`/`CF_SerializerWriteStream` for read and write interaction with anything inherited from the DayZ `Serializer` class.

Arkensor reviewed (changes requested):
Just a few minor comments to double-check before we can merge this.

Jacob-Mango committed:
Resolving comments
- Optimized Clamp calls by replacing with direct AND
- Documentation

Jacob-Mango committed:
Merge branch 'development' into feature/serializer

Jacob-Mango committed:
Docs

Jacob-Mango committed:
Note about ASCII
